drivers/virtio/virtio_guest_shm.c

```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Virtio over ivshmem front-end device driver
+ *
+ */
+
+#include <linux/delay.h>
+#include <linux/module.h>
+#include <linux/mutex.h>
+#include <linux/pci.h>
+#include <linux/dma-map-ops.h>
+#include <linux/memremap.h>
+#include <linux/virtio.h>
+#include <linux/virtio_config.h>
+#include <linux/virtio_ring.h>
+#include <linux/virtio_ids.h>
+#include <linux/virtio_pci.h>
+#include <linux/guest_shm.h>
+
+#include "virtio_shmem.h"
+
+#define DRV_NAME "virtio-guest-shm"
+
+struct virtio_guest_shm_data {
+	struct guest_shm_factory __iomem *fact;
+	struct guest_shm_control *ctrl;
+	unsigned int attach_list;
+};
+
+static void virtio_guest_shm_notify_peer(struct virtio_shmem_device *vi_dev, unsigned int vector)
+{
+	struct virtio_guest_shm_data *vi_data = vi_dev->priv;
+
+	writel(1 << vi_dev->peer_id, &vi_data->ctrl->notify);
+}
+
+static irqreturn_t virtio_guest_shm_early_irq_handler(struct virtio_shmem_device *vi_dev)
+{
+	struct virtio_guest_shm_data *vi_data = vi_dev->priv;
+	unsigned const attach_list = vi_data->ctrl->status >> GUEST_SHM_MAX_CLIENTS;
+	unsigned const attach_list_changes = attach_list ^ vi_data->attach_list;
+	unsigned int i;
+	irqreturn_t ret;
+
+	ret = IRQ_HANDLED;
+
+	for (i = 0; i < GUEST_SHM_MAX_CLIENTS; ++i) {
+		if (attach_list_changes & (1u << i))
+			dev_info(&vi_dev->pci_dev->dev, "client %2u %stached", i,
+				(attach_list & (1u << i)) ? "at" : "de");
+	}
+	vi_data->attach_list = attach_list;
+	return ret;
```

Comment:
Why to take separate variable ret for this. lets directly do return IRQ_HANDLED;

Reply:
OK. Changed.